New Histogram field mapper that supports percentiles aggregations. (#48580)
elastic/elasticsearch: docs/reference/mapping/types/histogram.asciidoc

* `histogram` fields do not support sorting.
========

[[histogram-uses]]
==== Uses

`histogram` fields are primarily intended for use with aggregations. To make it
more readily accessible for aggregations, `histogram` field data is stored as a
binary <<doc-values,doc values>> and not indexed. Its size in bytes is at most
`12 * numValues`, where `numValues` is the length of the provided arrays.

Because the data is not indexed, you only can use `histogram` fields for the
following aggregations and queries:

* <<search-aggregations-metrics-percentile-aggregation,percentiles>> aggregation
* <<search-aggregations-metrics-percentile-rank-aggregation,percentile ranks>> aggregation
* <<query-dsl-exists-query,exists>> query

We recommend you define the buckets in the `values` array based on the type of aggregation you intended to use.

[[mapping-types-histogram-building-histogram]]
==== Building a histogram

When using a histogram as part of an aggregation, the accuracy of the results will depend on how the
histogram was constructed. It is important to consider the percentiles aggregation mode that will be used
to build it. Some possibilities include:

> - For the <<search-aggregations-metrics-percentile-aggregation, T-Digest>> mode, histograms

Reviewer: Hmm, trying to tweak this a little to make it more explicit, so the user knows what the value/count fields do.

> - For the <<search-aggregations-metrics-percentile-aggregation, T-Digest>> mode, the `values` array represents the mean centroid positions and the `counts` array represents the number of values that are attributed to each centroid. If the algorithm has already started to approximate the percentiles, this inaccuracy is carried over in the histogram.

WDYT?

can be built by using the mean value of the centroids and the centroid's count. If the algorithm has already
started to approximate the percentiles, this inaccuracy is carried over in the histogram.

> - For the <<_hdr_histogram,High Dynamic Range (HDR)>> histogram mode, histograms

Reviewer: Similarly,

> - For the <<_hdr_histogram,High Dynamic Range (HDR)>> histogram mode, the `values` array represents fixed upper limits of each bucket interval, and the `counts` array represents the number of values that are attributed to each interval. This implementation maintains a fixed worse-case percentage error (specified as a number of significant digits), therefore the value used when generating the histogram would be the maximum accuracy you can achieve at aggregation time.

??

can be created by using the recorded values and the count at that value. This implementation maintains a fixed worse-case
percentage error (specified as a number of significant digits), therefore the value used when generating the histogram
would be the maximum accuracy you can achieve at aggregation time.

> 

Reviewer: Perhaps another sentence/paragraph at the end?

> The `histogram` field is "algorithm agnostic" and does not store data specific to either T-Digest or HDRHistogram.  While this means the field can technically be aggregated with either algorithm, in practice the user should chose one algorithm and index data in that manner (e.g. centroids for T-Digest or intervals for HDRHistogram) to ensure best accuracy.

Or something similar... trying to convey to the user that how they index the data is important and they should chose upfront.

[role="xpack"]
[testenv="basic"]
[[histogram]]
=== Histogram datatype
++++
<titleabbrev>Histogram</titleabbrev>
++++

A  field to store pre-aggregated numerical data representing a histogram.
This data is defined using two paired arrays:

* A `values` array of <<number, `double`>> numbers, representing the buckets for
the histogram. These values must be provided in ascending order.
* A corresponding `counts` array of <<number, `integer`>> numbers, representing how
many values fall into each bucket. These numbers must be positive or zero.

Because the elements in the `values` array correspond to the elements in the
same position of the `count` array, these two arrays must have the same length.

[IMPORTANT]
========
* A `histogram` field can only store a single pair of `values` and `count` arrays
per document. Nested arrays are not supported.
* `histogram` fields do not support sorting.
========

[[histogram-uses]]
==== Uses

`histogram` fields are primarily intended for use with aggregations. To make it
more readily accessible for aggregations, `histogram` field data is stored as a
binary <<doc-values,doc values>> and not indexed. Its size in bytes is at most

> `12 * numValues`, where `numValues` is the length of the provided arrays.

Reviewer 2: I think it's actually 13 since vints can take up to 5 bytes.

[role="xpack"]
[testenv="basic"]
[[histogram]]
=== Histogram datatype
++++
<titleabbrev>Histogram</titleabbrev>
++++

A `histogram` field stores pre-aggregated numerical data represented as a histogram. These histograms
are defined by a set of pairs, one being the value representing the bucket and the count of that value.

> THe values should be defined accordingly to the type of aggregation this field will be used for.

Reviewer 3: ```suggestion
The values should be defined accordingly to the type of aggregation this field will be used for.
```

Histograms can be used for aggregations, in particular <<search-aggregations-metrics-percentile-aggregation, percentiles>>
and <<search-aggregations-metrics-percentile-rank-aggregation, percentile ranks>> aggregations are supported.

Histograms are indexed by defining two arrays, one containing the `values` (or buckets) for the histogram and the other containing

> the `counts` for such values. Those two arrays must have the same length.

Reviewer 3: This section is great, but I think it's missing some crucial information:

- The `values` array stores `double` floats
- The `counts` array stores `long` integers
- The field data is stored as doc values and not indexed
- Because it's not indexed, it can't be used for most queries

I also think some of this section can get repetitive and needs some chunking. I left a suggestion below, but don't feel obligated to use it.

```
A  field to store pre-aggregated numerical data representing a histogram.
This data is defined using two paired arrays:

* A `values` array of <<number,`double`>> numbers, representing the buckets for
  the histogram. These values must be provided in ascending order.
* A corresponding `counts` array of  <<number,`long`>> numbers, representing how
  many values fall into each bucket.

Because the elements in the `values` array correspond to the elements in the
same position of the `count` array, these two arrays must have the same length.

[[histogram-uses]]
==== Uses

`histogram` fields are primarily intended for use with aggregations. To make it
more readily accessible for aggregations, `histogram` field data is stored as a
binary <<doc-values,doc values>> and not indexed.

Because the data is not indexed, you only can use `histogram` fields for the
following aggregations and queries:

* <<search-aggregations-metrics-percentile-aggregation,percentiles>> aggregation
* <<search-aggregations-metrics-percentile-rank-aggregation,percentile ranks>> aggregation
* <<query-dsl-exists-query,exists>> query

We recommend you define the buckets in the `values` array based on the type of aggregation you intended to use.
```

> 

Reviewer 3: Not strictly needed for MVP, but it might be nice to add some text for context and split up the index creation and document indexing snippets.

For example:

```
[[histogram-ex]]
==== Examples

The following <<indices-create-index, create index>> API request creates a new index with two field mappings:

* `my_histogram`, a `histogram` field used to store percentile data
* `my_text`, a `keyword` field used to store a title for the histogram

[ INSERT CREATE INDEX SNIPPET ]
...

The following <<docs-index_,index>> API requests store pre-aggregated for two histograms: `histogram_`` and `histogram_2`.

[ INSERT DOC INDEX SNIPPET ]
...
```

Providing an example use case for the data may also be helpful. For example, the histograms could represent load time, similar to the percentile aggs docs. Not required for MVP though.

[source,console]
--------------------------------------------------
PUT my_index
{
  "mappings": {
    "properties": {
      "my_histogram": {
        "type": "histogram",
      },
      "my_text" : {
        "type" : "keyword"
      }
    }
  }
}

PUT my_index/_doc/1
{
  "my_text" : "histogram_1",
  "my_histogram" : {
      "values" : [0.1, 0.2, 0.3, 0.4, 0.5], <1>
      "counts" : [3, 7, 23, 12, 6] <2>
   }
}

PUT my_index/_doc/2
{
  "my_text" : "histogram_2",
  "my_histogram" : {
      "values" : [0.1, 0.25, 0.35, 0.4, 0.45, 0.5], <1>
      "counts" : [8, 17, 8, 7, 6, 2] <2>
   }
}
--------------------------------------------------

<1> Values for each bucket. Values in the array are treated as doubles and must be given in

> increasing order. In case if TDDigest histograms this value represents the mean value.

Reviewer 3: ```suggestion
increasing order. For<<search-aggregations-metrics-percentile-aggregation-approximation,T-Digest>> histograms, this value represents the mean value.
```

> In case of HDR histograms this represents the value iterated to.

Reviewer 3: ```suggestion
For <<_hdr_histogram,High Dynamic Range (HDR)>> histograms, this represents the iterated value.
```

<2> Count for each bucket. Values in the arrays are treated as integers and must be positive or zero.
negative values will be rejected. The relation between a bucket and a count is given by the position in the array.

NOTE: `histogram` fields only support single-valued fields. Multi-valued fields
will be rejected.

NOTE: `histogram` fields with null field are ignored. Fields with empty `values` and `counts`
arrays are treated as null.

> NOTE: `rank_feature` fields do not support querying or sorting. They may

Reviewer 3: ```suggestion
NOTE: `histogram` fields do not support querying or sorting. They may
```

> only be used within <<query-dsl-exists-query, exists>> queries.

Reviewer 3: This is pretty important information. I included this content in my earlier suggestion, but otherwise you may want to move it above the example and condense it into a single `[IMPORTANT]` admonition.

Internally, each document's histogram is encoded as a binary
doc value. Its size in bytes is at most `12 * numValues`,

> where `numValues` is the length of the provided arrays.

Reviewer 3: It may be helpful to move this information up higher.

[[mapping-types-histogram-building-histogram]]
==== Building a histogram

> When using a histogram as part of an aggregation, the accuracy of the results will depend on how the

Reviewer 3: ```suggestion
When using a histogram as part of an aggregation, the accuracy of the results depends on how the
```

> histogram was constructed. It is important to consider the percentiles aggregation mode that will be used

Reviewer 3: ```suggestion
histogram was constructed. It is important to consider the percentiles aggregation mode that used
```

> in order to build it. Some possibilities are:

Reviewer 3: ```suggestion
to build it. Some possibilities include:
```

- For the <<search-aggregations-metrics-percentile-aggregation, T-Digest>> mode, histograms
can be built by using the mean value of the centroids and the centroid's count. If the algorithm has already
started to approximate the percentiles, this inaccuracy is carried over in the histogram.

> - For the https://github.com/HdrHistogram/HdrHistogram[HDR Histogram] (High Dynamic Range Histogram) method, histograms

Reviewer 3: It may be better to link to our HDR docs here: `<<_hdr_histogram,HDR>> (High Dynamic Range)...`
```suggestion
- For the https://github.com/HdrHistogram/HdrHistogram[HDR] (High Dynamic Range) histogram method, histograms
```
